### ISSUE TITLE
Support precedent rebind / changing the precedents in builtin library

### ISSUE THREAD

**Issue by HuStmpHrrr** [agda/agda#3574]
per https://github.com/agda/agda-stdlib/issues/624, the usable precedent range is too short, partly because builtin library has set precedents for a number of important notations.

I guess if we want to make the range larger, it either requires changing the builtin library, or Agda should allow changing the precedents of symbols that already exist.

**Comment by andreasabel** [comment]
I made a concrete suggestion for reassignment of fixities in #1346, suggesting a syntax like
```agda
open Product using (_×_) renaming (_,_ to infixr 5 _∷_)
```

**Comment by guillaumebrunerie** [comment]
Wouldn’t it be simpler to not introduce any new syntax but simply allow rebinding of fixities?
For instance:
```agda
open Product using (_×_) renaming (_,_ to _∷_)
infixr 5 _∷_
```

**Comment by nad** [comment]
> Wouldn’t it be simpler to not introduce any new syntax but simply allow rebinding of fixities?

Are you suggesting that it should be possible to change the fixity of an operator midway through a file?

**Comment by nad** [comment]
> I guess if we want to make the range larger, it either requires changing the builtin library, or Agda should allow changing the precedents of symbols that already exist.

Or we could switch to a different notion of precedence that is not so strongly tied to the natural numbers (preferably in a backwards-compatible way). @UlfNorell and I discussed one possibility in ["Parsing Mixfix Operators"](http://www.cse.chalmers.se/~nad/publications/danielsson-norell-mixfix.html), but other designs are possible.

**Comment by guillaumebrunerie** [comment]
> > Wouldn’t it be simpler to not introduce any new syntax but simply allow rebinding of fixities?
> 
> Are you suggesting that it should be possible to change the fixity of an operator midway through a file?

Yes. It’s probably not a very good practice but I see no strong reason for why it should be forbidden. Although maybe that the parser/fixity solver assumes that the fixity is the same everywhere?

**Comment by HuStmpHrrr** [comment]
> > I guess if we want to make the range larger, it either requires changing the builtin library, or Agda should allow changing the precedents of symbols that already exist.
> 
> Or we could switch to a different notion of precedence that is not so strongly tied to the natural numbers (preferably in a backwards-compatible way). @UlfNorell and I discussed one possibility in ["Parsing Mixfix Operators"](http://www.cse.chalmers.se/~nad/publications/danielsson-norell-mixfix.html), but other designs are possible.

another possibility is to support fixities at levels of rational numbers (arbitrary precision)? this is clearly backward-compatible.

**Comment by HuStmpHrrr** [comment]
> > > Wouldn’t it be simpler to not introduce any new syntax but simply allow rebinding of fixities?
> > 
> > 
> > Are you suggesting that it should be possible to change the fixity of an operator midway through a file?
> 
> Yes. It’s probably not a very good practice but I see no strong reason for why it should be forbidden. Although maybe that the parser/fixity solver assumes that the fixity is the same everywhere?

this would make the code extremely difficult to read and libraries very hard to maintain, if this feature is misused.

**Comment by guillaumebrunerie** [comment]
Well, essentially every feature can make the code difficult to read and the libraries hard to maintain if it is misused :)
But if it is only used in the use case considered here, for rebinding fixities of imported functions, then it should be fine.

**Comment by nad** [comment]
> > Are you suggesting that it should be possible to change the fixity of an operator midway through a file?

> Yes.

I don't think we should turn `infixr` and friends into imperative commands.

Furthermore, at the moment you can give the fixity of an operator after its declaration, so such a change might break code.

**Comment by guillaumebrunerie** [comment]
Isn’t it strange that the following code passes, but fails to parse if we comment the last line?
```agda
postulate
  A : Set
  _+_ : A → A → A
  _≡_ : A → A → Set

test : (a : A) → a + a ≡ a + a
test a = ?

infix 30 _+_
```
If the fixity declaration is much further down, it could probably get very confusing.
And I don’t see the problem with having them being "imperative commands". Other kinds of declarations, like declaring/defining identifiers, opening modules, various pragmas, and so on, are imperative.

**Comment by nad** [comment]
> it should be fine.

> it could probably get very confusing.

:)

I would prefer if the fixity declaration had to, say, precede the declaration of the identifier, but generally speaking I am hesitant towards introducing breaking changes for minor things like this.

> Other kinds of declarations, like declaring/defining identifiers, opening modules, various pragmas, and so on, are imperative.

You can't do things like removing a name so that it is out of scope, or changing the meaning of an identifier. (You can introduce a new identifier with the same unqualified name in a different scope.)

**Comment by gallais** [comment]
Fixed by 9c91d2f7bbb767f348618b6ca29fe68c66d19619
